src/lib.rs

```diff
@@ -1500,61 +1407,58 @@ impl EGraph {
         Ok(())
     }
 
-    fn process_command(
-        &mut self,
-        command: Command,
-        show_egglog_only: bool,
-    ) -> Result<Vec<ResolvedNCommand>, Error> {
-        let program = desugar::desugar_program(vec![command], &mut self.parser, self.seminaive)?;
+    fn process_command(&mut self, command: Command) -> Result<Vec<ResolvedNCommand>, Error> {
+        let mut program = self.resolve_command(command)?;
 
-        let mut program = self.typecheck_program(&program)?;
-
-        if !show_egglog_only {
-            program = remove_globals::remove_globals(program, &mut self.parser.symbol_gen);
-
-            for command in &program {
-                self.names.check_shadowing(command)?;
-            }
+        program = remove_globals::remove_globals(program, &mut self.parser.symbol_gen);
+        for command in &program {
+            self.names.check_shadowing(command)?;
         }
 
         Ok(program)
     }
 
+    fn resolve_command(&mut self, command: Command) -> Result<Vec<ResolvedNCommand>, Error> {
+        let program = desugar::desugar_program(vec![command], &mut self.parser, self.seminaive)?;
+        Ok(self.typecheck_program(&program)?)
+    }
+
     /// Run a program, represented as an AST.
     /// Return a list of messages.
-    pub fn run_program(&mut self, program: Vec<Command>) -> Result<Vec<String>, Error> {
-        let show_egglog_only = self.run_mode.show_egglog();
+    pub fn run_program(&mut self, program: Vec<Command>) -> Result<Vec<CommandOutput>, Error> {
+        let mut outputs = Vec::new();
         for command in program {
             // Important to process each command individually
             // because push and pop create new scopes
-            for processed in self.process_command(command, show_egglog_only)? {
-                if show_egglog_only {
-                    // In show_egglog mode, we still need to run scope-related commands (Push/Pop) to make
-                    // the program well-scoped.
-                    match &processed {
-                        ResolvedNCommand::Push(..) | ResolvedNCommand::Pop(..) => {
-                            self.run_command(processed.clone())?;
-                        }
-                        _ => {}
-                    };
-                    self.print_msg(processed.to_command().to_string());
```

Comment:
Previously, no-messages is a property of the e-graph, so run_program would respect that. With this change, there's no way for the user to turn off the messages when they call `run_program`, e.g., via methods in prelude? Is this desired?

Reply:
Hm, I mean now by default calling run program won't produce outputs, only return structured data, so if you don't format the results it won't ever been turned into a string. So this is similar to no messages? 

And logging in independent of course.